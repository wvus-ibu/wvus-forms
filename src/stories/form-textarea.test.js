import React, { Component } from "react";
import renderer from "react-test-renderer";
import { shallow, mount, render } from "enzyme";
import sinon from "sinon";

import {
  TextAreaForm,
  TextAreaFormTooMuchText,
  TextAreaFormNoConfig
} from "./shared/textarea-control-form";
import {
  formMethodsUnTouched,
  formMethodsWithSuccess,
  formMethodsWithError,
  formMethodsOptional
} from "./shared/props-mocks";
import { WVUSForm, Message, validateRequired, validateMax } from "../index";

const Form = WVUSForm(TextAreaForm);
const FormNoConfig = WVUSForm(TextAreaFormNoConfig);
const FormTooMuchText = WVUSForm(TextAreaFormTooMuchText);

/**
 * SNAPSHOTS
 */

test("Snapshot:  Textarea control renders correctly with no selection", () => {
  const tree = renderer.create(<Form />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("Snapshot:  Textarea control renders correctly with too much text", () => {
  const tree = renderer.create(<FormTooMuchText />).toJSON();
  expect(tree).toMatchSnapshot();
});

/**
 * Interactive - Enzyme tests
 */
describe("SelectControl", function() {
  const expectedFieldsState = {
    message: {
      value: "",
      isValid: false,
      secondInteraction: false,
      errorMessage: "",
      validators: [],
      optional: false
    }
  };

  it("should save initial field state", () => {
    const wrapper = mount(<FormNoConfig />);
    const actual = wrapper.state("fields");
    expect(actual).toEqual(expectedFieldsState);
  });

  it("getFieldState should return fieldState", () => {
    const wrapper = mount(<FormNoConfig />);
    const expected = expectedFieldsState.message;
    const actual = wrapper.getNode().getFieldState("message");
    expect(actual).toEqual(expected);
  });

  it("should call handleValueChange if field has changed", function() {
    sinon.spy(FormNoConfig.prototype, "handleValueChange");
    const wrapper = mount(<FormNoConfig />);
    wrapper.find("textarea").simulate("change", {
      target: { name: "message", value: "Some new message" }
    });
    expect(FormNoConfig.prototype.handleValueChange.calledOnce).toEqual(true);
  });

  it("should call handleBlur if field is blurred", function() {
    sinon.spy(FormNoConfig.prototype, "handleBlur");
    const wrapper = mount(<FormNoConfig />);
    wrapper.find("textarea").simulate("blur");
    expect(FormNoConfig.prototype.handleBlur.calledOnce).toEqual(true);
  });

  it("should show required star if required field", function() {
    const props = { formMethods: formMethodsUnTouched };
    const wrapper = mount(<TextAreaFormNoConfig {...props} />);
    expect(wrapper.contains(<sup>*</sup>)).toBe(true);
  });

  it("should NOT show required star if optional field", function() {
    const props = { formMethods: formMethodsOptional };
    const wrapper = mount(<TextAreaFormNoConfig {...props} />);
    expect(wrapper.contains(<sup>*</sup>)).toBe(false);
  });

  it("should have correct classes if optional field", function() {
    const props = {
      formMethods: formMethodsOptional
    };
    const wrapper = mount(<TextAreaFormNoConfig {...props} />);
    expect(wrapper.find(".has-error")).toHaveLength(0);
    expect(wrapper.find(".has-success")).toHaveLength(0);
  });

  it("should have correct classes if has success", function() {
    const props = {
      formMethods: formMethodsWithSuccess
    };
    const wrapper = mount(<TextAreaFormNoConfig {...props} />);
    expect(wrapper.find(".has-success")).toHaveLength(1);
    expect(wrapper.find(".has-error")).toHaveLength(0);
  });

  it("should have correct classes if has error", function() {
    const props = {
      formMethods: formMethodsWithError
    };
    const wrapper = mount(<TextAreaFormNoConfig {...props} />);
    expect(wrapper.find(".has-error")).toHaveLength(1);
    expect(wrapper.find(".has-success")).toHaveLength(0);
  });

  it("should show error Message if has error", function() {
    const props = {
      formMethods: formMethodsWithError
    };
    const wrapper = mount(<TextAreaFormNoConfig {...props} />);
    const errorMessage = props.formMethods.getFieldState().errorMessage;
    expect(
      wrapper.contains(
        <Message
          visible={true}
          showError={true}
          showSuccess={false}
          message={errorMessage}
        />
      )
    ).toBe(true);
  });

  it("should show success Message if has success", function() {
    const props = {
      formMethods: formMethodsWithSuccess
    };
    const wrapper = mount(<TextAreaFormNoConfig {...props} />);
    expect(
      wrapper.contains(
        <Message visible={true} showError={false} showSuccess={true} />
      )
    ).toBe(true);
  });
});
