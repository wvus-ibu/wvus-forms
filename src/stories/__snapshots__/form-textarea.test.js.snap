// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot:  Textarea control renders correctly with no selection 1`] = `
<div
  className=""
>
  <form>
    <div
      className="form-group has-icon "
    >
      <label
        className="control-label"
        htmlFor="message"
      >
        Message
         
        <sup>
          *
        </sup>
      </label>
      <div
        className="form-control-wrapper"
      >
        <div
          className="text-area-wrapper"
        >
          <textarea
            className="form-control"
            id="message"
            name="message"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Type your message here"
            rows={6}
            value=""
          />
          <span
            className="message character-count"
          >
            255
          </span>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`Snapshot:  Textarea control renders correctly with too much text 1`] = `
<div
  className=""
>
  <form>
    <div
      className="form-group has-icon has-error "
    >
      <label
        className="control-label"
        htmlFor="message"
      >
        Message
         
        <sup>
          *
        </sup>
      </label>
      <div
        className="form-control-wrapper"
      >
        <div
          className="text-area-wrapper"
        >
          <textarea
            className="form-control"
            id="message"
            name="message"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Type your message here"
            rows={6}
            value="Bacon ipsum dolor amet cow kevin pork chop turducken ball tip chuck sausage strip steak tri-tip alcatra. Shankle landjaeger tenderloin short ribs burgdoggen alcatra corned beef salami ground round porchetta turducken. Short loin rump pork belly, ham hock beef ribs ribeye salami corned beef turducken jerky fatback tenderloin porchetta swine pork loin. Brisket chuck short ribs leberkas shoulder ham hock. Frankfurter buffalo turducken short loin jerky drumstick, andouille bacon. Beef ribs leberkas short loin, andouille ball tip ribeye burgdoggen short ribs. Pork loin venison meatloaf hamburger t-bone. Pork loin pork swine t-bone flank brisket. Beef ribs spare ribs salami drumstick landjaeger ball tip kielbasa pork belly pork meatball short ribs brisket short loin boudin jerky. Meatloaf salami prosciutto flank picanha, beef ribs capicola tenderloin pork chop jowl shoulder swine. Chicken jowl salami, leberkas biltong capicola hamburger filet mignon ball tip ham swine pig pork brisket. Fatback pork belly chicken pastrami leberkas, ribeye porchetta doner sirloin biltong cupim."
          />
          <span
            className="message character-count"
          >
            -831
          </span>
        </div>
        <span>
          <i
            className="fa fa-exclamation-circle"
          />
          <span
            className="message"
            data-field-validation-message={true}
          >
            
          </span>
        </span>
      </div>
    </div>
  </form>
</div>
`;
